File: 227_Basic Calculator II.java
// 在遇到下一個operator時處理前一個的運算
// Solution: Optimised Approach without the stack
// Time Complexity: O(n)
// Space Complexity: O(1)

class Solution {
    public int calculate(String s) {
        if (s == null || s.length() == 0) return 0;

        int cur = 0, last = 0, res = 0;
        Character operator = '+';

        for (int i = 0; i < s.length(); i++) {
            Character c = s.charAt(i);

            // number
            if (Character.isDigit(c)) cur = cur*10 + (c - '0');
            // operator || process the last digit
            if ((!Character.isDigit(c) && !Character.isWhitespace(c)) || (i == s.length() - 1)) {
                if (operator == '+' || operator == '-') {
                    res += last;
                    last = (operator == '+') ? cur : 0-cur;
                }
                else if (operator == '*') {
                    last *= cur;
                }
                else if (operator == '/') {
                    last /= cur;
                }

                operator = c;
                cur = 0;
            }
        }
        
        return (res + last);
    }
}

File: 227_Basic Calculator II.py
class Solution:
    def calculate(self, s: str) -> int:
        # Stack: */ have higher priority than +-, so process them first
        # For -, just convert the sign for next num
        stack = []
        prevOper = '+'
        num = 0

        for i, c in enumerate(s):
            # all possible cases: digit+-*/
            if c.isdigit(): num = num * 10 + int(c)
            # process the previous operation
            if i == len(s) - 1 or c in "+-*/":
                if prevOper == '*': stack.append(num * stack.pop())
                # Python: -3//2 = -2 -> （向负无穷方向取整）cannot use directly -> int(-3/2)
                elif prevOper == '/': 
                    top = stack.pop()
                    if top < 0: stack.append(int(top / num))
                    else: stack.append(top // num)
                elif prevOper == '-': stack.append(-num)
                elif prevOper == '+': stack.append(num)

                prevOper = c
                num = 0

        return sum(stack)